CIE-5464 - Backup and Restore for tool based installations (#3928)
SoftwareAG/c8y-docs: content/edge-kubernetes/edge-operations-bundle/backup-and-recovery.md

---
weight: 10

> title: Backup and Recovery

Reviewer: ```suggestion
title: Backup and recovery
```

layout: redirect
---

> # Cumulocity Edge on K3s — Backup and Restore Guide

Reviewer: This heading can be removed. It is obvious from the title, just a repetition.

Author: Corrected the heading to make it more sensinble since we may get another generic backup and restore doc as well.

Reviewer: Then you should better change the title accordingly and remove this heading. Or, if you have more than one section below "backup and restore", then make this a level 3 heading (###) and all other headings need to be changed to level 4 (####). This is the general structure we follow.

Author: done

Reviewer 2: On that note, this doc does in fact need the generic (and brief section) mentioned in CIE-5464 .

This section heading would probably scan nicer if it was just "For c8yedge installations" rather than "Backup and restore guide for c8yedge installations"; less redundancy with the overall title, and will contrast nicely with another heading like "For Edge on self-managed kubernetes".

This runbook describes how to capture and restore a **{{< product-c8y-iot >}} Edge** deployment running on **K3s** installed via the `c8yedge` tool. Follow the numbered steps to create a consistent backup, reinstall the same Edge version, and validate the restored environment.

---

> ### Step 1 - Create a backup 

Reviewer: Please add heading IDs to all headings

Author: Any pointers for that on how you want me to name the IDs ?

Reviewer: Seems I overlooked your comment. It is always the entire heading text, all lower case, words combined with a hyphen. 
Step 1 - Create a backup
#step-1---create-a-backup

Author: I have added IDs as suggested however i am not able to see them on the UI, Usualy it comes on the right side for quick navigation.

<img width="1640" height="783" alt="image" src="https://github.com/user-attachments/assets/069537ca-4505-4b7e-8277-2a8debfaaff2" />

> #### Step 1 - Create a backup {#step-1---create-a-backup}

Reviewer 2: "Create a backup" implies discrete periodic copies - but that doesn't count if you take the decision to have your live data on a reliable NAS.

So I think this sections needs to be even more vague, like "Data to preserve", which is simply pointing out the directories that need to be immune to disaster. Then suggesting either backups or having these as being on some kind of reliable storage like a NAS.

Back up the required directories using your preferred backup strategy.

* `/var/lib/rancher/k3s` - always required

> * `/datahub` - only if Cumulocity DataHub is deployed (contains DataHub datalake content)

Reviewer: ```suggestion
* `/datahub` - only if {{< product-c8y-iot >}} DataHub is deployed (contains DataHub datalake content)
```

> Ensure the backup is stored securely. Where applicable, retain checksum or integrity metadata to allow validation prior to recovery.

Reviewer 2: I think this line is too [teaching grandmother to suck eggs](https://en.wikipedia.org/wiki/Teaching_grandmother_to_suck_eggs) - _how_ to do backups/securing of data, isn't really our business.

{{< c8y-admon-caution >}}
Installing a different Edge version on top of a restored data set is unsupported and may fail the upgrade guard rails.
{{< /c8y-admon-caution >}}

---

### Step 3 - Restore the data directories
Restore the backup so that the directories land in their original locations, preserving paths, ownership, and permissions. Omitting directories or restoring to incorrect locations can corrupt the cluster.

Confirm the directories exist and contain the expected ownership:

```shell
ls -ld /var/lib/rancher/k3s
# If Cumulocity DataHub was deployed, also verify:
[ -d /datahub ] && ls -ld /datahub
```

---

### Step 4 - Reinstall the matching Edge release
Re-run the installer with the **exact version** captured in the backup:

```shell
sudo c8yedge install --version <previous_version>
```

Or use the offline alternative if you are in an airgapped environment:

```shell
sudo c8yedge install -s c8yedge.tar
```

For more details on installation options, see [Installing with the c8yedge tool](/edge-kubernetes/installing-edge-on-k8/#install-edge-with-c8yedge-cli) and [Install Edge in an airgapped environment](/edge-kubernetes/installing-edge-on-k8/#install-edge-airgapped).

> The installer identifies the Cumulocity Edge CR if it is being restored and waits for the operator to reconcile successfully before exiting.

Reviewer: ```suggestion
The installer identifies the {{< product-c8y-iot >}} Edge CR if it is being restored and waits for the operator to reconcile successfully before exiting.
```

Watch for the following success messages:

```shell
...

> 2025-12-17T11:59:47Z	Cumulocity Edge update is complete in 3m14s (running version 2025.0.X)

Reviewer: ```suggestion
2025-12-17T11:59:47Z	{{< product-c8y-iot >}} Edge update is complete in 3m14s (running version 2025.0.X)
```

> The installer identifies the {{< product-c8y-iot >}} Edge CR if it is being restored and waits for the operator to reconcile successfully before exiting.

Reviewer 2: Bit too technical (for a `c8yedge` user) - what we're trying to tell them at a higher level is that installer identifies that Cumulocity Edge is already installed, and will wait for recovery to complete.

---
weight: 10
title: Backup and Recovery
layout: redirect
---

> # Cumulocity Edge on K3s — Backup and Restore Guide

Reviewer 2: Either here (or maybe even the section title?) let's make it clear that we're only really supporting if you installed with `c8yedge`. Somebody might be using self-managed k3s, and end up in a situation that nobody was expecting.

Author: Mentioned it as part of the doc itself 

`This runbook describes how to capture and restore a **{{< product-c8y-iot >}} Edge** deployment running on **K3s** installed via the `c8yedge` tool. Follow the numbered steps to create a consistent backup, reinstall the same Edge version, and validate the restored environment.`

> ### For self managed c8yedge installations {#for-self-managed-c8yedge-installations}

Reviewer 2: ```suggestion
### For installations on self-managed kubernetes clusters {#for-self-managed-c8yedge-installations}
```
Don't want the name of the tool 'c8yedge' anywhere near this section title! This is more consistent with our wording in the 'Installation' chapter.

> ### For installations on self-managed kubernetes clusters {#for-self-managed-c8yedge-installations}

Reviewer: ```suggestion
### For installations on self-managed Kubernetes clusters {#for-self-managed-c8yedge-installations}
```

This runbook describes how to capture and restore a **{{< product-c8y-iot >}} Edge** deployment running on **K3s** (installed via the `c8yedge` tool). Follow the numbered steps to create a consistent backup, reinstall the same Edge version, and validate the restored environment.

{{< c8y-admon-important >}}

> Always back up `/var/lib/rancher/k3s` (and `/datahub` if Cumulocity DataHub is deployed) together and restore them to their original paths, ownership, and permissions. Mixing versions or omitting directories can corrupt the cluster.

Reviewer 2: This seems to be a slightly reworded alias to the text in `Step 1 - Create a backup`. Combine?

Author: Reframed the doc to make the instruction flow cleaner and avoid duplication

> Once a storage class is configured in the Edge custom resource (CR), it cannot be changed. Ensure you select the appropriate storage class during initial installation, as this setting is immutable after Edge is deployed.

Reviewer 2: May as well put in the detail here - the field is `storageClassName`.

{{< /c8y-admon-important >}}

---

### Step 1 - Create a backup 
Back up the required directories using your preferred backup strategy. Ensure you capture:

* `/var/lib/rancher/k3s` - always required
* `/datahub` - only if Cumulocity DataHub is deployed (contains DataHub datalake content)

Preserve the original paths, ownership, and permissions during the backup process. Ensure the backup is stored securely. Where applicable, retain checksum or integrity metadata to allow validation prior to recovery.

---

### Step 2 - Prepare the restore target

> 1. Install the same operating system (or compatible base image) that originally hosted Edge.

Reviewer 2: What's "base image" referring to here? I normally take this as referring to OCI images, but I guess this is something VM related ... ?

Author: Vocabulary from Appliance days, Corrected it in the latest commit.

---
weight: 10
title: Backup and Recovery
layout: redirect
---

# Cumulocity Edge on K3s — Backup and Restore Guide

This runbook describes how to capture and restore a **Cumulocity Edge** deployment that runs on **K3s** (installed via the `c8yedge` tool). Follow the numbered steps to collect a consistent snapshot, reinstall the same Edge version, and validate the restored environment.

{{< c8y-admon-important >}}

> Always back up `/var/lib/rancher/k3s` and `/datahub` together and restore them to their original paths, ownership, and permissions. Mixing versions or omitting directories can corrupt the cluster.

Reviewer 2: Mention that `/datahub` is optional here, if you've not installed it you'll not have such a thing on disk.

Author: Noted - Done

{{< /c8y-admon-important >}}

---

### Step 1 - Understand what must be protected
* `/var/lib/rancher/k3s`: cluster state, objects, certificates, and etcd/sqlite internals
* `/datahub`: {{< product-c8y-iot >}} DataHub (datalake) content, if provisioned on the node

---

### Step 2 - Prepare the Edge node
Stopping workloads is optional but prevents partial state during the archive.

```shell
systemctl stop k3s
```

Ensure only one backup job runs at a time and confirm the filesystem has enough free space to hold the archive.

> ---

Reviewer 2: I don't think we can recommend this - it's not a reasonable thing for somebody to do in a production system every time they take a backup.

_If_ we end up having to recommend this (if we can't get a running system reliably backed up and restored in our tests), then we should be more definite i.e. "You must do this". Having something optional like this is going to lead to anxiety.

Author: removed

> ### Step 3 - Create a consolidated backup 

Reviewer 2: There are many different ways of doing backups
* Periodic tar-balling
* LVM/BTRFS snapshots
* Using NFS
* Anything I've not thought of
so I don't think we should to dictate specific commands, especially as tar-balling isn't the best of these, I think we should be more generic in our advice.

It's good to call out which directories matter, as we have done, but beyond that go vague
e.g. "preserve <these directories> (maybe a periodic backup, maybe just having them on NAS)" and "make sure that <these directories> are restored"

Author: This makes the doc even simpler, removing the steps as suggested.

```shell
tar -czvf edge-backup.tar.gz /var/lib/rancher/k3s /datahub
```

Move `edge-backup.tar.gz` to offline storage or to your backup repository. Retain checksum information if you wish to valdiate prior to recovery.

---

### Step 4 - Prepare the restore target
1. Install the same operating system (or compatible base image) that originally hosted Edge.
2. Make sure no prior K3s installation or Edge data exists on the target disk.
3. Transfer `edge-backup.tar.gz` to the target node.

{{< c8y-admon-caution >}}
Installing a different Edge or K3s version on top of a restored data set is unsupported and may fail the upgrade guard rails.
{{< /c8y-admon-caution >}}

---

### Step 5 - Restore the data directories
Extract the archive to the root of the filesystem so that both directories land in their original locations:

```shell
tar -xzvf edge-backup.tar.gz -C /
```

Confirm the directories exist and contain the expected ownership:

```
/var/lib/rancher/k3s
/datahub
```

---

### Step 6 - Reinstall the matching Edge release
Re-run the installer with the **exact version** captured in the backup:

```shell
c8yedge install --version <previous_version>

> ```

Reviewer 2: Or the offline alternative

Author: Noted - Done

The installer reattaches the restored cluster state and Edge-specific configuration.

---

> ### Step 7 - Verify health

Reviewer 2: I'm _hoping_ this step is unnecessary - the `c8yedge` tool should only exit once the whole system is healthy, meaning they don't need to know how to read the kubectl output.

Author: refined it with the current logging we have added which clearly says recovery successful, that will replace this doc to let customer know what message to watchout for success status.
